Remora.Discord.Extensions/Embeds/EmbedBuilder.cs

```diff
@@ -0,0 +1,384 @@
+﻿//
+//  EmbedBuilder.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Drawing;
+using System.Linq;
+using Remora.Discord.API;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Objects;
+using Remora.Discord.Extensions.Errors;
+using Remora.Rest.Core;
+using Remora.Results;
+
+namespace Remora.Discord.Extensions.Embeds
+{
+    /// <summary>
+    /// Provides utilities for building an embed.
+    /// </summary>
+    public class EmbedBuilder
+    {
+        /// <summary>
+        /// Gets or sets the title of the embed.
+        /// </summary>
+        [MaxLength(Constants.MaxTitleLength)]
+        public string Title { get; set; } = string.Empty;
+
+        /// <summary>
+        /// Gets the type of embed. Only <see cref="EmbedType.Rich"/> is available for embeds created by bots.
+        /// </summary>
+        public EmbedType Type { get; } = EmbedType.Rich;
+
+        /// <summary>
+        /// Gets or sets the description of the embed.
+        /// </summary>
+        [MaxLength(Constants.MaxDescriptionLength)]
+        public string Description { get; set; } = string.Empty;
+
+        /// <summary>
+        /// Gets or sets the url of the embed.
+        /// </summary>
+        [Url]
+        public string Url { get; set; } = string.Empty;
```

Comment:
If this an Url, should it not be defined as a `Uri`?

Reply:
This field is a string because [`IEmbed.Url`](https://github.com/Nihlus/Remora.Discord/blob/master/Backend/Remora.Discord.API.Abstractions/API/Objects/Messages/IEmbed.cs#L55) is a string. The `UrlAttribute` ensures that it is a properly formatted url.

Reply:
Gotcha.